### PR TITLE
Add instance normalization

### PR DESCRIPTION
**Pull request by crcrpar**
This adds instance normalization #5229 and I'll close #5309 because that is a bit too old and I gave up to manage running statistics that are not used in the original paper.

### CODE REVIEW

File: chainer/functions/normalization/instance_normalization.py [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+
+    Args:
+        x (:class:`~chainer.Variable` or :class:`numpy.ndarray` or \
+        :class:`cupy.ndarray`): Batch tensors.
+            First dimension of this value must be the size of minibatch and
+            second dimension must be the number of channels.
+            Moreover, this value must have one or more following dimensions,
+            such as height and width.
+        gamma (~chainer.Variable): Scaling parameter.
+        beta (~chainer.Variable): Shifting parameter.
```
Comment (hvy) [inline chainer/functions/normalization/instance_normalization.py]: How about explaining the expected shapes of `gamma` and `beta`?

File: chainer/functions/normalization/instance_normalization.py [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+
+from chainer.functions.array import reshape
+from chainer.functions.array import tile
+from chainer.functions.normalization import batch_normalization
+
+
+def instance_normalization(x, gamma, beta, eps=2e-5):
+    """Instance normalization function.
+
+    This function implements a "instance normalization"
```
Comment (hvy) [inline chainer/functions/normalization/instance_normalization.py]: You can remove the `a` and possible simplify it as follows.
```suggestion
    This function implements instance normalization
```

File: chainer/functions/normalization/instance_normalization.py [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+    Returns:
+        ~chainer.Variable: The output variable which has the same shape
+        as :math:`x`.
+
+    See: `Instance Normalization: The Missing Ingredient for Fast Stylization
+           <https://arxiv.org/abs/1607.08022>`_
+
+    """
+    if x.ndim <= 2:
+        raise ValueError('Input dimension must be grater than 2, '
```
Comment (hvy) [inline chainer/functions/normalization/instance_normalization.py]: Ops, typo.
```suggestion
        raise ValueError('Input dimension must be greater than 2, '
```

File: chainer/functions/normalization/instance_normalization.py [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+                         'including batch size dimension '
+                         '(first dimension).')
+
+    batch_size, channels = x.shape[:2]
+    original_shape = x.shape
+    x = reshape.reshape(x, (1, batch_size * channels) + x.shape[2:])
```
Comment (hvy) [inline chainer/functions/normalization/instance_normalization.py]: A minor but how about accessing `x.shape` only once when accessing `original_shape` and then use `original_shape` in the following accesses.

File: chainer/links/normalization/instance_normalization.py [hunk 5]
```diff
@@ -0,0 +1,140 @@
...
+        return initializers.generate_array(
+            initializer, size, self.xp, dtype=self._dtype)
+
+    def forward(self, x):
+        """forward(self, x, finetune=False)
+
+        Args:
+            x (Variable): Input variable.
+
+        """
```
Comment (hvy) [inline chainer/links/normalization/instance_normalization.py]: Let's make the description slightly more verbose.

File: chainer/links/normalization/instance_normalization.py [hunk 6]
```diff
@@ -0,0 +1,140 @@
...
+
+    Args:
+        size (int): Size (or shape) of channel dimensions.
+        eps (float): Epsilon value for numerical stability.
+        dtype (numpy.dtype): Type to use in computing.
+        use_gamma (bool): If ``True``, use scaling parameter. Otherwise, use
+            unit(1) which makes no effect.
+        use_beta (bool): If ``True``, use shifting parameter. Otherwise, use
+            unit(0) which makes no effect.
+
```
Comment (hvy) [inline chainer/links/normalization/instance_normalization.py]: Another minor but for consistency.
```suggestion
```

File: tests/chainer_tests/links_tests/normalization_tests/test_instance_normalization.py [hunk 7]
```diff
@@ -0,0 +1,172 @@
...
+
+        self.x = numpy.random.uniform(-1, 1, self.shape).astype(self.dtype)
+        self.gy = numpy.random.uniform(-1, 1, self.shape).astype(self.dtype)
+
+        self.check_forward_options = {'atol': 1e-4, 'rtol': 1e-3}
+        self.check_backward_options = {'atol': 1e-3, 'rtol': 1e-2}
+
+    def check_forward(self, x_data):
+        y = self.link(x_data)
+        # self.assertEqual(y.data.dtype, self.dtype)
```
Comment (hvy) [inline tests/chainer_tests/links_tests/normalization_tests/test_instance_normalization.py]: Is this comment intentional?

File: chainer/functions/normalization/instance_normalization.py [hunk 8]
```diff
@@ -0,0 +1,49 @@
...
+    gamma = tile.tile(gamma, batch_size)
+    beta = tile.tile(beta, batch_size)
```
Comment (hvy) [inline chainer/functions/normalization/instance_normalization.py]: I think it's nicer to check for the dimensions and throw appropriate errors here, rather than from batch norm but what do you think?

File: chainer/links/normalization/instance_normalization.py [hunk 9]
```diff
@@ -0,0 +1,140 @@
...
+from chainer import variable
+
+
+class InstanceNormalization(link.Link):
+
+    """Instance normalization layer on outputs of linear or convolution functions.
+
+    This link wraps the :func:`~chainer.functions.instance_normalization`.
+    Instance normalization is very close to batch normalization but different
+    in that this normalizes each samples in mini batch by its mean and standard
```
Comment (hvy) [inline chainer/links/normalization/instance_normalization.py]: Grammar and minor cosmetic for consistency.
```suggestion
    in that this normalizes each samples in a mini-batch by its mean and standard
```

File: chainer/links/normalization/instance_normalization.py [hunk 10]
```diff
@@ -0,0 +1,140 @@
...
+    deviation even if in testing mode. Also note that this normalization only
+    work on inputs whose dimensions are greater than 2.
```
Comment (hvy) [inline chainer/links/normalization/instance_normalization.py]: Sorry another minor.
```suggestion
    works on inputs whose dimensions are greater than 2.
```

File: chainer/links/normalization/instance_normalization.py [hunk 11]
```diff
@@ -0,0 +1,140 @@
...
+    def forward(self, x):
+        """forward(self, x, finetune=False)
+
+        Args:
+            x (Variable): Input variable.
+
+        """
+
+        return instance_normalization.instance_normalization(
+            x, self.gamma, self.beta, self.eps)
```
Comment (hvy) [inline chainer/links/normalization/instance_normalization.py]: Is it okay to pass `self.{gamma,beta}` as `None` to the function?

File: chainer/links/normalization/instance_normalization.py [hunk 12]
```diff
@@ -0,0 +1,140 @@
...
+            self.beta.initialize(shape)
+
+    def _init_array(self, initializer, default_value, size):
+        if initializer is None:
+            initializer = default_value
+        initializer = initializers._get_initializer(initializer)
+        return initializers.generate_array(
+            initializer, size, self.xp, dtype=self._dtype)
+
+    def forward(self, x):
```
Comment (hvy) [inline chainer/links/normalization/instance_normalization.py]: Don't we need to check if we're training or testing and use/update the running values accordingly?

**Comment by hvy** [comment]
Thanks for this PR! I left some initial comments, some slight nitpicks but would you mind taking a look?

File: chainer/functions/normalization/instance_normalization.py [hunk 13]
```diff
@@ -0,0 +1,49 @@
...
+
+    with warnings.catch_warnings():
+        warnings.simplefilter("ignore")
```
[the remark this answers is not in the document]
Comment (crcrpar) [reply chainer/functions/normalization/instance_normalization.py]: Do you mean add an explanation to the code or explain why this filter exists here?

If latter, this is because inputs for batch_norm.batch_norm have the batchsize of 1 and batchnorm functions seem to warn such inputs.

File: tests/chainer_tests/functions_tests/normalization_tests/test_instance_normalization.py [hunk 14]
```diff
@@ -0,0 +1,112 @@
...
+
+        def func(*args):
+            return functions.instance_normalization(*args)
+
+        y = func(*args)
+        # self.assertEqual(y.dtype, self.dtype)
+        assert y.dtype == self.dtype
+
+        # Verify that implementation using batch normalization is correct
+        y_simple_gn = _simple_instance_normalization(*self.args, xp)
```
[the remark this answers is not in the document]
Comment (crcrpar) [reply tests/chainer_tests/functions_tests/normalization_tests/test_instance_normalization.py]: I didn't care this. Thanks.
> This style of passing xp after an unpacked argument is not supported in Python 3.4 or less.

File: chainer/functions/normalization/instance_normalization.py [hunk 15]
```diff
@@ -0,0 +1,49 @@
...
+
+def instance_normalization(x, gamma, beta, eps=2e-5):
+    """Instance normalization function.
+
+    This function implements a "instance normalization"
+    which normalizes each sample by its mean and standard deviation.
+
+    Args:
+        x (:class:`~chainer.Variable` or :class:`numpy.ndarray` or \
+        :class:`cupy.ndarray`): Batch tensors.
```
[the remark this answers is not in the document]
Comment (crcrpar) [reply chainer/functions/normalization/instance_normalization.py]: Sure.

File: tests/chainer_tests/links_tests/normalization_tests/test_instance_normalization.py (hunk 7 above)
Comment (crcrpar) [reply tests/chainer_tests/links_tests/normalization_tests/test_instance_normalization.py]: No, it's my fault.

File: chainer/links/normalization/instance_normalization.py (hunk 11 above)
Comment (crcrpar) [reply chainer/links/normalization/instance_normalization.py]: It doesn't seem to OK and I'll resolve this.

File: chainer/functions/normalization/instance_normalization.py (hunk 8 above)
Comment (crcrpar) [reply chainer/functions/normalization/instance_normalization.py]: I agree with you. Is it OK to leave scaling and shifting to batch_norm function like this?

**Comment by crcrpar** [comment]
- add a feature of tracking moving stats.
- use_gamma and use_beta are `False` by default.
    - updates should be done in instance_norm function.
- copy arguments of BatchNorm link and replace argument `axis` with `track_avg_stats`

**Comment by crcrpar** [comment]
double backward is so unstable...
It frequently returns nan.

File: chainer/functions/normalization/instance_normalization.py (hunk 15 above)
[the remark this answers is not in the document]
Comment (toslunar) [reply chainer/functions/normalization/instance_normalization.py]: Could you also fix `Args` section of `L.InstanceNormalization`?

**Comment by hvy** [comment]
Can you find out under what test parameter configurations?

**Comment by crcrpar** [comment]
Sorry.

The parameters are dtype=numpy.float16, use_cudnn=True, and use_fast_batchnormalization=True.

**Comment by crcrpar** [comment]
note.
As of now, I'm struggling to fix double backward when
1. dtype is float16, use_cudnn is 'always', and cudnn_fast_batch_normalization is `True`. In this case, double backward is unstable so that frequently returns nan.
2. use_chainerx is `True`. In this case, double backward does not work due to the broadcast error. I guess the error occurs around gamma and beta broadcast and reshape.

File: tests/chainer_tests/functions_tests/normalization_tests/test_instance_normalization.py [hunk 16]
```diff
@@ -152,16 +153,16 @@ def setUp(self):
     def generate_inputs(self):
         shape, dtype = self.shape, self.dtype
         x = numpy.random.uniform(-1, 1, shape).astype(dtype)
-        gamma = numpy.random.uniform(0.5, 1, shape[1]).astype(dtype)
+        gamma = numpy.random.uniform(-1, 1, shape[1]).astype(dtype)
         beta = numpy.random.uniform(-1, 1, shape[1]).astype(dtype)
         mean = numpy.random.uniform(-1, 1, shape[1]).astype(dtype)
-        var = numpy.random.uniform(-1, 1, shape[1]).astype(dtype)
+        var = numpy.random.uniform(.5, 1, shape[1]).astype(dtype)
```
Comment (toslunar) [inline tests/chainer_tests/functions_tests/normalization_tests/test_instance_normalization.py]: Nice catch!

**Comment by crcrpar** [comment]
It seems that I cannot send running_mean and running_var to a device...

Isn't this enough?

https://github.com/chainer/chainer/blob/f8a32fac37d95276acca9b9d3d89cf0004014675/tests/chainer_tests/functions_tests/normalization_tests/test_instance_normalization.py#L103-L114
image

File: chainer/links/normalization/instance_normalization.py [hunk 17]
```diff
@@ -0,0 +1,247 @@
...
+           <https://arxiv.org/abs/1607.08022>`_
+
+    .. seealso::
+       :func:`~chainer.functions.instance_normalization`,
+
+    Attributes:
+        gamma (:class:`~chainer.Variable`): Scaling parameter.
+        beta (:class:`~chainer.Variable`): Shifting parameter.
+        avg_mean (:ref:`ndarray`): Population mean.
+        avg_var (:ref:`ndarray): Population variance.
```
Comment (toslunar) [inline chainer/links/normalization/instance_normalization.py]:
```suggestion
        avg_var (:ref:`ndarray`): Population variance.
```

**Comment by hvy** [comment]
It seems like Travis is failing.

**Comment by crcrpar** [comment]
The test failure is the same as this https://github.com/chainer/chainer/pull/5847#issuecomment-477914814.

I can reproduce this error but I cannot this when I call `F.instance_normalization` in other scripts or ipython.

File: chainer/functions/normalization/instance_normalization.py [hunk 18]
```diff
@@ -0,0 +1,170 @@
...
+        x = tile.tile(x, reps)
+    elif xp is chainerx:
+        x = xp.concatenate((x,) * reps)
+    else:
+        x = xp.tile(x, reps)
+    return x
+
+
+def instance_normalization(x, gamma, beta, **kwargs):
+    """instance_normalization(x, gamma, beta, *, running_mean=None,
```
Comment (toslunar) [inline chainer/functions/normalization/instance_normalization.py]:
```suggestion
    """instance_normalization(x, gamma, beta, *, running_mean=None, \
```

File: chainer/functions/normalization/instance_normalization.py [hunk 19]
```diff
@@ -0,0 +1,170 @@
...
+            Scaling parameter of input with the shape of (C,).
+        decay (float): Decay rate of moving average. It is used during
+            training.
+
+    Returns:
+        :class:`~chainer.Variable`: The output variable which has the same
+        shape as :math:`x`.
+
+    See: `Instance Normalization: The Missing Ingredient for Fast Stylization
+           <https://arxiv.org/abs/1607.08022>`_
```
Comment (toslunar) [inline chainer/functions/normalization/instance_normalization.py]:
```suggestion
    <https://arxiv.org/abs/1607.08022>`_
```

File: chainer/links/normalization/instance_normalization.py [hunk 20]
```diff
@@ -0,0 +1,247 @@
...
+            is ``1``.
+        initial_beta: Initializer of the shifting parameter. The default value
+            is ``0``.
+        initial_avg_mean: Initializer of the moving average of population mean.
+            The default value is ``0``.
+        initial_avg_var: Initializer of the moving average of population
+            variance. The default value is ``1``.
+
+    See: `Instance Normalization: The Missing Ingredient for Fast Stylization
+           <https://arxiv.org/abs/1607.08022>`_
```
Comment (toslunar) [inline chainer/links/normalization/instance_normalization.py]: Ditto.
